File: backend/app/schemas/proforma.py
"""Proforma Invoice schemas.

Independent replica of the Quotation schemas (app/schemas/sales.py),
renamed to Proforma Invoice. Reuses SalesLineItemRequest for line items.
"""
from datetime import date, datetime
from typing import List, Optional
from uuid import UUID
from pydantic import BaseModel, ConfigDict

from app.schemas.sales import SalesLineItemRequest


class ProformaInvoiceCreateRequest(BaseModel):
    customer_id: UUID
    proforma_date: date
    valid_until: Optional[date] = None
    sold_to_customer_id: Optional[UUID] = None
    bill_to_customer_id: Optional[UUID] = None
    ship_to_customer_id: Optional[UUID] = None
    notes: Optional[str] = None
    terms_conditions: Optional[str] = None
    status: str = "draft"
    items: List[SalesLineItemRequest]


class ProformaInvoiceStatusRequest(BaseModel):
    status: str


class ProformaInvoiceResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    proforma_number: str
    customer_id: UUID
    proforma_date: Optional[date]
    valid_until: Optional[date]
    status: str
    subtotal: int
    total_discount: int
    total_taxable_amount: int
    total_cgst: int
    total_sgst: int
    total_igst: int
    total_gst: int
    total_amount: int
    notes: Optional[str]
    terms_conditions: Optional[str]
    created_at: Optional[datetime]


class ProformaInvoicesListResponse(BaseModel):
    items: List[ProformaInvoiceResponse]
    total: int
    page: int
    page_size: int
    has_more: bool

File: frontend/src/api/proforma.ts
/**
 * Proforma Invoice API Client
 *
 * Independent replica of the Quotation client (api/sales.ts) renamed to
 * Proforma Invoice. Talks to /api/v2/proforma-invoices.
 */

import { apiClient } from './client';
import type { SalesLineItem } from './sales';

export interface CreateProformaInvoicePayload {
  customer_id: string;
  proforma_date: string;
  valid_until?: string;
  sold_to_customer_id?: string;
  bill_to_customer_id?: string;
  ship_to_customer_id?: string;
  notes?: string;
  terms_conditions?: string;
  status?: string;
  items: SalesLineItem[];
}

export type UpdateProformaInvoicePayload = Partial<CreateProformaInvoicePayload>;

export interface ProformaInvoice {
  id: string;
  proforma_number: string;
  customer_id: string;
  proforma_date: string;
  valid_until?: string;
  status: string;
  sold_to_customer_id?: string;
  bill_to_customer_id?: string;
  ship_to_customer_id?: string;
  subtotal: number;
  total_discount: number;
  total_taxable_amount: number;
  total_cgst: number;
  total_sgst: number;
  total_igst: number;
  total_gst: number;
  total_amount: number;
  notes?: string;
  terms_conditions?: string;
  created_at: string;
}

class ProformaApiClient {
  async listProformaInvoices(
    status?: string,
    page = 1,
    page_size = 20,
    options?: { archived_only?: boolean; include_archived?: boolean }
  ) {
    const params: Record<string, string | number | boolean> = { page, page_size };
    if (status) params.status = status;
    if (options?.archived_only) params.archived_only = true;
    if (options?.include_archived) params.include_archived = true;
    return apiClient.get<{ items: ProformaInvoice[]; total: number }>('/api/v2/proforma-invoices', { params });
  }

  async getProformaInvoice(id: string) {
    return apiClient.get<ProformaInvoice>(`/api/v2/proforma-invoices/${id}`);
  }

  async createProformaInvoice(payload: CreateProformaInvoicePayload) {
    return apiClient.post<ProformaInvoice>('/api/v2/proforma-invoices', payload);
  }

  async updateProformaInvoice(id: string, payload: UpdateProformaInvoicePayload) {
    return apiClient.put<ProformaInvoice>(`/api/v2/proforma-invoices/${id}`, payload);
  }

  async updateProformaInvoiceStatus(id: string, status: string) {
    return apiClient.patch<ProformaInvoice>(`/api/v2/proforma-invoices/${id}/status`, { status });
  }

  async archiveProformaInvoice(id: string) {
    return apiClient.patch<ProformaInvoice>(`/api/v2/proforma-invoices/${id}/archive`, {});
  }

  async restoreProformaInvoice(id: string) {
    return apiClient.patch<ProformaInvoice>(`/api/v2/proforma-invoices/${id}/restore`, {});
  }

  async downloadProformaInvoicePdf(id: string) {
    return apiClient.get(`/api/v2/proforma-invoices/${id}/pdf`, {
      responseType: 'blob',
    });
  }

  async sendProformaInvoiceEmail(id: string, email?: string) {
    return apiClient.post(`/api/v2/proforma-invoices/${id}/send-email`, { email });
  }
}

export const proformaApi = new ProformaApiClient();
